### PR TITLE
Add MusicUpdateFunction field to RTD

### PR DESCRIPTION
**Pull request by frutiemax**
The circus ride type has a special behaviour, so it was needed to add a MusicUpdateFunction field to the RTD. Right now, all RTDs use DefaultMusicUpdate, Circus uses CircusMusicUpdate.

### CODE REVIEW

File: src/openrct2/ride/Ride.cpp [hunk 1]
```diff
@@ -1880,20 +1880,8 @@ static void CircusMusicUpdate(Ride* ride)
...
+void DefaultMusicUpdate(Ride* ride)
 {
-    if (ride->type == RIDE_TYPE_CIRCUS)
-    {
-        CircusMusicUpdate(ride);
-        return;
-    }
-
-    const auto& rtd = ride->GetRideTypeDescriptor();
-    if (!rtd.HasFlag(RIDE_TYPE_FLAG_MUSIC_ON_DEFAULT) && !rtd.HasFlag(RIDE_TYPE_FLAG_ALLOW_MUSIC))
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Ride.cpp:1892]: To me it looks like there are 3 settings Default, Circus and None.

Comment (frutiemax) [reply src/openrct2/ride/Ride.cpp:1892]: It was an error from me, this check is necessary.

File: src/openrct2/ride/Ride.cpp [hunk 2]
```diff
@@ -1926,6 +1914,15 @@ static void ride_music_update(Ride* ride)
     OpenRCT2::RideAudio::UpdateMusicInstance(*ride, rideCoords, sampleRate);
 }
 
+static void ride_music_update(Ride* ride)
+{
+    const auto& rtd = ride->GetRideTypeDescriptor();
+
+    if (!rtd.HasFlag(RIDE_TYPE_FLAG_MUSIC_ON_DEFAULT) && !rtd.HasFlag(RIDE_TYPE_FLAG_ALLOW_MUSIC))
+        return;
+    rtd.MusicUpdateFunction(ride);
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Ride.cpp:1923]: When the RTD is converted to an object this property will become an enum (probably) so perhaps it would make sense to do:
```cpp
switch(rtd.MusicUpdateFunction)
{
    case MusicUpdateFunction::Circus:
      CircusMusicUpdate(ride);
      break;
    case MusicUpdateFunction::Default:
       DefaultMusicUpdate(ride);
       break;
}```

**Review by duncanspumpkin — COMMENTED** [review]
You've accidentally added a testdata file.

File: src/openrct2/ride/Ride.cpp (hunk 2 above)
Comment (frutiemax) [reply src/openrct2/ride/Ride.cpp:1923]: The switch case adds a very small performance overhead compared to the function pointer I believe... Also the TrackPaintFunction will check against a string when importing/exporting to json, I would do the same for the MusicUpdateFunction. It's not the same with the OperatingMode as the value is checked at multiple places in the code.

Comment (duncanspumpkin) [reply src/openrct2/ride/Ride.cpp:1923]: I would seriously doubt it would make a difference but I'm not overly fussed for now just think it might simplify things further down the line.

**Review by duncanspumpkin — APPROVED** [review]

File: src/openrct2/ride/Ride.cpp [hunk 3]
```diff
@@ -1926,6 +1914,15 @@ static void ride_music_update(Ride* ride)
     OpenRCT2::RideAudio::UpdateMusicInstance(*ride, rideCoords, sampleRate);
 }
 
+static void ride_music_update(Ride* ride)
```
Comment (tupaschoal) [inline src/openrct2/ride/Ride.cpp:1917]:
```suggestion
static void ride_music_update(Ride* ride)
```
New functions should be titlecase

**Review by tupaschoal — COMMENTED** [review]
Wouldn't it be more efficient memory-wise if we set just the ones that didn't use the `DefaultMusicUpdate`?

**Comment by frutiemax** [comment]
> Wouldn't it be more efficient memory-wise if we set just the ones that didn't use the `DefaultMusicUpdate`?

It would still require a function pointer for the field even if unset, i.e. it would take the same amount of memory. Also, if you do it like that, you would need to check if that field is null/not null. I prefer to avoid making checks for null and call the correct function directly.

File: src/openrct2/ride/RideData.h [hunk 4]
```diff
@@ -378,7 +379,7 @@ constexpr const RideTypeDescriptor DummyRTD =
     SET_FIELD(BonusValue, 0),
     SET_FIELD(ColourPresets, DEFAULT_FLAT_RIDE_COLOUR_PRESET),
     SET_FIELD(ColourPreview, { static_cast<uint32_t>(SPR_NONE), static_cast<uint32_t>(SPR_NONE) }),
-    SET_FIELD(ColourKey, RideColourKey::Ride)
+    SET_FIELD(ColourKey, RideColourKey::Ride),
```
Comment (tupaschoal) [inline src/openrct2/ride/RideData.h:382]: Unrelated change?

File: src/openrct2/ride/coaster/meta/LIMLaunchedRollerCoaster.h [hunk 5]
```diff
@@ -52,6 +52,5 @@ constexpr const RideTypeDescriptor LIMLaunchedRollerCoasterRTD =
     )),
     SET_FIELD(ColourPreview, { SPR_RIDE_DESIGN_PREVIEW_LIM_LAUNCHED_ROLLER_COASTER_TRACK, SPR_RIDE_DESIGN_PREVIEW_LIM_LAUNCHED_ROLLER_COASTER_SUPPORTS }),
     SET_FIELD(ColourKey, RideColourKey::Ride),
-
```
Comment (tupaschoal) [inline src/openrct2/ride/coaster/meta/LIMLaunchedRollerCoaster.h:55]: Unrelated change

**Review by tupaschoal — APPROVED** [review]

**Comment by tupaschoal** [comment]
There is a merge conflict that needs sorting out for this to be merged @frutiemax

File: src/openrct2/ride/shops/meta/CashMachine.h [hunk 6]
```diff
@@ -46,6 +46,7 @@ constexpr const RideTypeDescriptor CashMachineRTD =
     SET_FIELD(ColourPresets, DEFAULT_FLAT_RIDE_COLOUR_PRESET),
     SET_FIELD(ColourPreview, { 0, 0 }),
     SET_FIELD(ColourKey, RideColourKey::CashMachine),
+    SET_FIELD(MusicUpdateFunction, DefaultMusicUpdate),
```
Comment (Broxzier) [inline src/openrct2/ride/shops/meta/CashMachine.h:49]: Can't these be dropped? Same goes for the 6 other ones.

Comment (frutiemax) [reply src/openrct2/ride/shops/meta/CashMachine.h:49]: You cant, not with the SET_FIELD method.

Comment (Broxzier) [reply src/openrct2/ride/shops/meta/CashMachine.h:49]: But `MusicUpdateFunction` is initialized with `DefaultMusicUpdate` by default, right?
